Guide for adding Canvas badges (#333)
scroll-tech/scroll-documentation: src/content/docs/en/developers/guides/canvas-badge-integration.mdx

---
section: developers
date: Last Modified
title: "Canvas & Badge Integration Guide"
lang: "en"
permalink: "developers/guides/canvas-badge-integration"
excerpt: "This guide contains instructions on how to issue your own badge on Scroll Canvas."
---

import Aside from "../../../../../components/Aside.astro"
import BadgeDetails from "../_images/badgeDetails.jpg"
import BadgeLevels from "../_images/badgeLevels.png"
import ClickToZoom from "../../../../../components/ClickToZoom.astro"
import Columns from "../../../../../components/Columns.astro"
import ToggleElement from "../../../../../components/ToggleElement.astro"

We are thrilled to have you join us in building unique badges on Scroll Canvas, a product designed for ecosystem projects to interact with users in a more personal way on-chain. 

## Background

### Canvas

> Each Canvas is a `Profile` smart contract, minted by the users through the `ProfileRegistry` contract on Scroll. Canvas is not transferrable and is unique to each wallet.

Reviewer: This page should link to https://github.com/scroll-tech/canvas-contracts somewhere, otherwise it's very hard for devs to reference what these contracts are.

Canvas is an open onchain profile database of user identities and achievements ([check out dashboard built by community on Dune](https://dune.com/soodoo/scroll-marks)).

### Badge

> **Badges** are attestations of identities, achievements and traits verified through the [Ethereum Attestation Service](https://scroll.easscan.org/learn/scroll), issued **permissionlessly by different ecosystem projects**. Badges are wallet-bound and non-transferable. **Badges facilitate interactions between ecosystem projects and users**. 

Reviewer: ```suggestion
**Badges** are attestations of identities, achievements and traits issued through the [Ethereum Attestation Service](https://scroll.easscan.org/learn/scroll) **permissionlessly by different ecosystem projects**. Badges are wallet-bound and non-transferable. **Badges facilitate interactions between ecosystem projects and users**. 
```

Developers can issue badges in three methods:

| Issuance Method       | **Description**                                       | **Example**                                                                       |
| ---                   | ---                                                   | ---                                                                                  |
| `Permissionless`      | Badge checks eligibility based on smart contract.     | Badges attest to the completion of an on-chain transaction or holding of an NFT. |
| `Backend-authorized`  | Badge checks eligibility based on the issuer’s API.   | Badges attest to the completion of off-chain actions or a certain allow list. |

> | `Gifted`              | Badges checks eligibility based on the issuer’s API and **automatically** sends the badges to users’ canvas. There is no minting required for users to display the badge.  | Badges attest to the ownership or paid membership on other platforms / chains. |

Reviewer: ```suggestion
| `Gifted`              | Badge checks eligibility based on the issuer’s API and **automatically** sends the badges to users’ canvas. There is no minting required for users to display the badge.  | Badges attest to the ownership or paid membership on other platforms / chains. |
```

Developers can design badges in two ways:

| Badge Structure           | Description                                                           |
| ---                       | ---                                                                   |
| **`Singleton badges`**    | static attestations based on a certain user action/attribute          |

> | **`Leveled badges`**      | dynamic attestations based on the progression of the users’ onchain   |

Reviewer: This sentence is incomplete

## Before you Start

Badge issuance and deployment of the contract is fully permissionless. Please follow the steps below to issue a Badge for your project.

> <Aside type="note" title="Join Developer Support Channel">
> For Canvas and badge questions, please join [Scroll dev support channel](https://discord.com/invite/scroll)

Reviewer: ```suggestion
<Aside type="note" title="Join the Developer Support Channel">
For Canvas and badge questions, please join the [Scroll dev support channel](https://discord.com/invite/scroll)
```

</Aside>

## Step 1 `Design`: Badge Design Guidelines 

<Columns>
    <Fragment slot="1"><ClickToZoom src={BadgeDetails} alt="BadgeDetails" caption="Example of badge detail page"/></Fragment>
    <Fragment slot="2"><ClickToZoom src={BadgeLevels} alt="BadgeLevels" caption="Example of a Leveled Badge" /></Fragment>
</Columns>

### Visual Format Request:

> - Maximum resolution: 480px x 480px

Reviewer: Is this "minimum"? Since it's smaller than "optimal"?

- Optimal resolution: 600px x 600px
- File size: Under 300KB

### Decide on Badge Details

| Badge Details     | Additional Info |
| ---               | --- |
| Name              | Name your badge to be *descriptive and creative* |
| Description       | Brief description of the *purpose* of this badge and *who is eligible* |
| Structure         | *One single attribute*<br/>  *Multiple Badges with levels* trigged by different actions and traits |

> | Category          | *Achievement*: tasks you have done. e.g. Scroll Origin NFTs, tasks completion badge <br/> *Identities*: who you are. e.g. Ethereum year, GItcoin passport, NFT community badge |

Reviewer: ```suggestion
| Category          | *Achievement*: tasks you have done. e.g. Scroll Origin NFTs, tasks completion badge <br/> *Identities*: who you are. e.g. Ethereum year, Gitcoin passport, NFT community badge |
```

        3. **contentType**: `data/json`
        4. **responseSchema**
            
            ```jsx
            // eligible
            {
            	code: 1,
            	message: "success",
            	eligibility: true
            }
            
            // not eligible
            {
            	code: 0,
            	message: "why the recipient is not eligible",
            	eligibility: false
            }
            
            // error
            // http code is not 200
            ```
        
        ### Claim Badge
        
        1. **method**: GET
        2. **url**: `{{pathname}}/claim?badge={{badgeContractAddress}}&recipient={{walletAddress}}`
        3. **contentType**: `data/json`
        4. **responseSchema**
            
            ```jsx
            // success
            {
            	code: 1,
            	message: "success",
            	tx: ...
            }
            
            // failure
            {
            	code: 0,
            	message: "some msg",
            }
            
            // error
            // http code is not 200
            ```
    </ToggleElement>
    

> 4. **Access Readiness**: Ensure badge contract address implements `defaultTokenURI`, allowing retrieval of default display data (`name, image, description`) via `bytes32(0)` with informative name and description. 

Reviewer: ```suggestion
4. **Access Readiness**: Ensure your badge contract implements `ScrollBadgeDefaultURI`, allowing retrieval of default display data (`name, image, description, issuerName `) via `bytes32(0)` with informative name and description. 
```

5. **Set up support**: Prepare a channel for your badge holders can find you for support, get the url to join that channel ready.

## Step 3 `Test`: Sanity Check 

> 1. Once your badge has been deployed on Scroll, you can auto-check some of these requirements by running `yarn check-badge`.([link to check badge script](https://github.com/scroll-tech/canvas-contracts/blob/master/script/CheckBadge.s.sol))

Reviewer: ```suggestion
1. Once your badge has been deployed on Scroll, you can auto-check some (but not all) of these requirements by running `yarn check-badge` ([link to check badge script](https://github.com/scroll-tech/canvas-contracts/blob/master/script/CheckBadge.s.sol)).
```

    - If your badge minting transaction reverts, we recommend debugging using `cast`:

    ```bash
    cast run --rpc-url https://rpc.scroll.io [txhash]
    ```

    This call will simulate the transaction in a local environment and show you the call stack and revert reason.
    - For `badgeTokenURI` link returned, ensure correct cross-origin configuration
    <ToggleElement anchor="reference-cors">
    <div slot="title">Reference: CORS</div>

            ### CORS
            
            > Please ensure the provided API supports the cross-origin requirements.
            > 
            
            ### Sepolia
            
            - **localhost**: This allows for development and testing directly from your local machine environment.
            - **Domains containing the substring 'scroll'**: Any domain that includes the substring 'scroll' in its name is allowed to make requests.
            
            ### Mainnet
            
            - **`scroll.io` domain**: requests originating from **`scroll.io`** are permitted.
    </ToggleElement>

> 2. **Review the information on the badge detail page**: 

Reviewer: Where is "the badge detail page"?

Reviewer 2: we dont have link here, the section is asking issuer to check all the details they provide is correct on their preview page.
